packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/library/CopyLibraryImportHandler.java

```diff
@@ -0,0 +1,131 @@
+
+    public CopyLibraryImportHandler(ILibrarySearchService librarySearchService, ISemanticDataSearchService semanticDataSearchService, IResourceLoader resourceLoader, List<IEditingContextMigrationParticipantPredicate> migrationParticipantPredicates, ICollaborativeMessageService messageService, MeterRegistry meterRegistry) {
+        this.librarySearchService = Objects.requireNonNull(librarySearchService);
+        this.semanticDataSearchService = Objects.requireNonNull(semanticDataSearchService);
+        this.resourceLoader = Objects.requireNonNull(resourceLoader);
+        this.migrationParticipantPredicates = Objects.requireNonNull(migrationParticipantPredicates);
+        this.messageService = Objects.requireNonNull(messageService);
+        this.counter = Counter.builder(Monitoring.EVENT_HANDLER)
+                .tag(Monitoring.NAME, this.getClass().getSimpleName())
+                .register(meterRegistry);
+    }
+
+    @Override
+    public boolean canHandle(IEditingContext editingContext, IInput input) {
+        return input instanceof ImportLibrariesInput importLibrariesInput
+                && Objects.equals(importLibrariesInput.type(), "copy");
+    }
+
+    @Override
+    public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IInput input) {
+        this.counter.increment();
+
+        String message = this.messageService.invalidInput(input.getClass().getSimpleName(), ImportLibrariesInput.class.getSimpleName());
+        IPayload payload = new ErrorPayload(input.id(), message);
+        ChangeDescription changeDescription = new ChangeDescription(ChangeKind.NOTHING, editingContext.getId(), input);
+
+        if (input instanceof ImportLibrariesInput importLibrariesInput && Objects.equals(importLibrariesInput.type(), "copy")) {
+            if (editingContext instanceof EditingContext siriusWebEditingContext) {
+                List<Library> libraries = this.librarySearchService.findAllById(importLibrariesInput.libraryIds().stream().map(UUID::fromString).toList());
+
+                Set<SemanticData> semanticDataToCopy = new LinkedHashSet<>();
+                for (Library library : libraries) {
+                    this.semanticDataSearchService.findById(library.getSemanticData().getId())
+                        .ifPresent(semanticData -> {
+                            semanticDataToCopy.add(semanticData);
+                            semanticDataToCopy.addAll(this.semanticDataSearchService.findAllDependenciesRecursivelyById(semanticData.getId()));
+                        });
+                }
+
+                for (SemanticData semanticData: semanticDataToCopy) {
+                    for (Document document : semanticData.getDocuments()) {
+                        Optional<Resource> optResource = this.resourceLoader.toResource(siriusWebEditingContext.getDomain().getResourceSet(), document.getId().toString(), document.getName(), document.getContent(),
```

Comment:
I think that this can be simplified, I'm not convinced that we need to mess with the URI converter for example.

Reply:
Ok, here is my reasoning in case it helps you find a better solution:
- The documents of the library may contain proxies to other documents of the library (or dependencies of the library). We don't load the dependencies in any particular order, so there will be some proxies.
- If I load all the documents into resources with an URI based on the document's ID, I am sure the proxies can be resolved
- I can't keep the same ID because of the unicity of `document#id` in the database, so I have to change the ID of the resource to be able to persist it.